[WIP] rbd configurable worker threads

MahatiC opened the pull request:
This PR introduces refcounting for reference counting state machines and aims to support configurable worker threads (via "rbd op threads")
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

MahatiC commented:
Rebased onto master (Oct 8th). Fixes test failures.
@dillaman Please review.

dillaman commented on src/librbd/Utils.h:
```diff
@@ -7,10 +7,10 @@
 #include "include/rados/librados.hpp"
 #include "include/rbd_types.h"
 #include "include/Context.h"
+#include "common/RefCountedObj.h"
```
Nit: unused

MahatiC replied:
Done

dillaman commented on src/librbd/DeepCopyRequest.cc:
```diff
@@ -215,13 +215,18 @@ void DeepCopyRequest<I>::send_copy_object_map() {
   ceph_assert(m_snap_seqs->count(m_snap_id_end) > 0);
   librados::snap_t copy_snap_id = (*m_snap_seqs)[m_snap_id_end];
+
+  using klass = DeepCopyRequest<I>;
```
Nit: in general, I think it would be preferable to hide the reference counting to within the affected classes. For example, in ```ObjectMap::rollback```, it could wrap the provided ```Context``` with a helper in ```Utils.h``` that invokes the ```get``` on construction and ```put``` on destruction. That should prevent most of this churn throughout librbd.

MahatiC replied:
Ack. Perhaps a wraparound like this? - https://gist.github.com/MahatiC/95fc2133127c17d4389d20511c2c3f76 

I think this would simplify the code quite a bit, though it wouldn't be applicable to all of the changes.

dillaman replied:
Yeah -- I think something like that looks good. Perhaps rename the helper/wrapper to something including ref-counting (i.e. ```create_context_callback``` -> ```create_ref_counted_context```)

MahatiC commented:
@dillaman sure. Just to confirm my understanding of yaml. Is this what you mean? https://gist.github.com/MahatiC/f4e00fc5f54a693b4866b49aca41067d

I'm going to use rbd op threads value 8 for qa/suites/rbd unless there's a preference for something else.

dillaman commented:
> @dillaman sure. Just to confirm my understanding of yaml. Is this what you mean? https://gist.github.com/MahatiC/f4e00fc5f54a693b4866b49aca41067d

Yup, something like that. It just needs to be added to each sub-test directory under the RBD suite. 

> I'm going to use rbd op threads value 8 for qa/suites/rbd unless there's a preference for something else.

Sure. It also might be good to add a config override to ```src/test/librbd/test_main.cc``` and ```src/test/rbd_mirror/test_main.cc``` so that the unit tests pick up the change as well.

MahatiC commented:
> Sure. It also might be good to add a config override to `src/test/librbd/test_main.cc` and `src/test/rbd_mirror/test_main.cc` so that the unit tests pick up the change as well.

Done

MahatiC commented:
retest this please

dillaman commented:
@MahatiC I'm sure it's just the increased op work threads causing odd race conditions under the Jenkins environment since both tests seg faulted. You might need to run the tests locally in a loop to re-create: ```ctest -V -R run-rbd-unit-test```

MahatiC commented:
@dillaman I agree. Unfortunately even the looping doesn't reproduce the errors for me :\ Attaching a couple of test run outputs as well (avoiding run-rbd-unit-test output as it's too large to fit here).
I'm wondering if there's any other way - maybe Jenkins env has more info available on this? 
[DeepCopy_repeat.txt](https://github.com/ceph/ceph/files/2515060/DeepCopy_repeat.txt)
[unittest_librbd.txt](https://github.com/ceph/ceph/files/2515061/unittest_librbd.txt)

dillaman commented:
@MahatiC It cannot run in teuthology because it's not even passing the unit tests. I will try to make some time to look into it locally later this week.

MahatiC commented:
> @MahatiC It cannot run in teuthology because it's not even passing the unit tests. I will try to make some time to look into it locally later this week.

okay, thank you

MahatiC commented:
@dillaman do you have any update on this?

dillaman commented:
@MahatiC It crashes or deadlocks for me on every attempt to run the unittests:

```
[ RUN      ] TestMockDeepCopyRequest.SimpleCopy

Thread 1 "unittest_librbd" received signal SIGBUS, Bus error.
0x000055555627a58b in ThreadPool::PointerWQ<Context>::drain (this=0x61635f747865746e) at /home/jdillaman/ceph_wip/src/librbd/ImageCtx.cc:71
71	    op_work_queue->drain();
(gdb) bt
#0  0x000055555627a58b in ThreadPool::PointerWQ<Context>::drain (this=0x61635f747865746e) at /home/jdillaman/ceph_wip/src/librbd/ImageCtx.cc:71
#1  librbd::(anonymous namespace)::ThreadPoolSingleton::~ThreadPoolSingleton (this=0x555556d369b0, __in_chrg=<optimized out>) at /home/jdillaman/ceph_wip/src/librbd/ImageCtx.cc:71
#2  ceph::_any::op_func<librbd::(anonymous namespace)::ThreadPoolSingleton> (o=ceph::_any::op::destroy, p=0x555556d369b0) at /home/jdillaman/ceph_wip/src/include/any.h:80
#3  0x0000555555b614e8 in ceph::_any::base<ceph::immobile_any<576ul>, std::aligned_storage<576ul, 16ul>::type>::reset (this=0x555556d369a0) at /home/jdillaman/ceph_wip/src/common/RWLock.h:178
#4  ceph::_any::base<ceph::immobile_any<576>, std::aligned_storage<576, 16>::type>::construct<librbd::(anonymous namespace)::ThreadPoolSingleton, CephContext*&> (this=0x555556d369a0) at /home/jdillaman/ceph_wip/src/include/any.h:206
#5  ceph::_any::base<ceph::immobile_any<576>, std::aligned_storage<576, 16>::type>::base<librbd::(anonymous namespace)::ThreadPoolSingleton, CephContext*&> (this=0x555556d369a0) at /home/jdillaman/ceph_wip/src/include/any.h:319
#6  ceph::immobile_any<576>::immobile_any<librbd::(anonymous namespace)::ThreadPoolSingleton, CephContext*&> (this=0x555556d369a0) at /home/jdillaman/ceph_wip/src/include/any.h:545
#7  std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> >::pair<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&, 0, 1, std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&, 0, 1> (__tuple2=..., __tuple1=..., 
    this=0x555556d36970) at /usr/include/c++/8/tuple:1668
#8  std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> >::pair<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&, std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> (__second=..., __first=..., this=0x555556d36970)
    at /usr/include/c++/8/tuple:1657
#9  __gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > > >::construct<std::pair<std::pair<std::__cxx11::basic_string<char>, std::type_index> const, ceph::immobile_any<576> >, std::piecewise_construct_t const&, std::tuple<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&>, std::tuple<std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> > (this=<optimized out>, __p=0x555556d36970) at /usr/include/c++/8/ext/new_allocator.h:136
#10 std::allocator_traits<std::allocator<std::_Rb_tree_node<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > > > >::construct<std::pair<std::pair<std::__cxx11::basic_string<char>, std::type_index> const, ceph::immobile_any<576> >, std::piecewise_construct_t const&, std::tuple<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&>, std::tuple<std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> > (__a=..., __p=0x555556d36970) at /usr/include/c++/8/bits/alloc_traits.h:475
#11 std::_Rb_tree<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index>, std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> >, std::_Select1st<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > >, CephContext::associated_objs_cmp, std::allocator<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > > >::_M_construct_node<std::piecewise_construct_t const&, std::tuple<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&>, std::tuple<std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> > (this=<optimized out>, __node=0x555556d36950) at /usr/include/c++/8/bits/stl_tree.h:637
#12 std::_Rb_tree<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index>, std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> >, std::_Select1st<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > >, CephContext::associated_objs_cmp, std::allocator<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > > >::_M_create_node<std::piecewise_construct_t const&, std::tuple<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&>, std::tuple<std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> > (this=<optimized out>) at /usr/include/c++/8/bits/stl_tree.h:654
#13 std::_Rb_tree<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index>, std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> >, std::_Select1st<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > >, CephContext::associated_objs_cmp, std::allocator<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > > >::_M_emplace_hint_unique<std::piecewise_construct_t const&, std::tuple<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&>, std::tuple<std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> > (__pos=..., this=<optimized out>) at /usr/include/c++/8/bits/stl_tree.h:2414
#14 std::map<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index>, ceph::immobile_any<576>, CephContext::associated_objs_cmp, std::allocator<std::pair<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::type_index> const, ceph::immobile_any<576> > > >::emplace_hint<std::piecewise_construct_t const&, std::tuple<std::basic_string_view<char, std::char_traits<char> >&, std::type_index&>, std::tuple<std::in_place_type_t<librbd::(anonymous namespace)::ThreadPoolSingleton> const&, CephContext*&> > (__pos=
    {first = {first = "", second = {_M_target = 0x555556c776e0}}, second = {<ceph::_any::base<ceph::immobile_any<576>, std::aligned_storage<576, 16>::type>> = {static capacity = <optimized out>, func = 0x0, storage = {__data = '\000' <repeats 24 times>, "P_\313VUU\000\000p_\313VUU\000\000\220_\313VUU\000\000\000\304\312VUU", '\000' <repeats 26 times>, "\240*\317VUU\000\000\240*\317VUU\000\000\240*\317VUU\000\000\001\000\000\000\000\000\000\000\200\026\261VUU\000\000\240\311\312VUU\000\000p?\307VUU", '\000' <repeats 18 times>, "\020P\303VUU\000\000\020P\303VUU\000\000\200P\303VUU\000\000\002", '\000' <repeats 71 times>..., __align = {<No data fields>}}}, static capacity = 576}}, this=0x555556c776a8) at /usr/include/c++/8/bits/stl_map.h:607
#15 CephContext::lookup_or_create_singleton_object<librbd::(anonymous namespace)::ThreadPoolSingleton, CephContext*&> (drop_on_fork=false, Python Exception <class 'gdb.error'> value has been optimized out: 
name=, this=<optimized out>) at /home/jdillaman/ceph_wip/src/common/ceph_context.h:188
#16 librbd::ImageCtx::get_thread_pool_instance (cct=<optimized out>, thread_pool=0x7fffffffcf48, op_work_queue=0x555556cf9f40) at /home/jdillaman/ceph_wip/src/librbd/ImageCtx.cc:899
#17 0x00005555562854f6 in librbd::ImageCtx::ImageCtx(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*, librados::IoCtx&, bool) () at /home/jdillaman/ceph_wip/src/librbd/ImageCtx.cc:141
#18 0x00005555560dd929 in TestFixture::open_image(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, librbd::ImageCtx**) () at /usr/include/c++/8/bits/char_traits.h:320
#19 0x0000555555c227d9 in TestMockDeepCopyRequest::SetUp (this=0x555556aeeb70) at /home/jdillaman/ceph_wip/src/test/librbd/test_mock_DeepCopyRequest.cc:154
#20 0x00005555564a004a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (location=0x555556599999 "SetUp()", method=<optimized out>, object=0x555556aeeb70) at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest-internal-inl.h:803
#21 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object@entry=0x555556aeeb70, method=<optimized out>, location=location@entry=0x555556599999 "SetUp()") at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest.cc:2438
#22 0x0000555556496ee6 in testing::Test::Run (this=0x555556aeeb70) at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest.cc:2470
#23 0x0000555556497098 in testing::TestInfo::Run (this=0x555556b18a40) at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest.cc:2656
#24 0x0000555556497175 in testing::TestCase::Run (this=0x555556b08a70) at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest.cc:2774
#25 0x0000555556497694 in testing::internal::UnitTestImpl::RunAllTests() () at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest-internal-inl.h:586
#26 0x00005555564a052a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (location=0x55555659ab90 "auxiliary test code (environments or event listeners)", method=<optimized out>, object=0x555556b11730) at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest-internal-inl.h:803
#27 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x555556b11730, method=<optimized out>, location=location@entry=0x55555659ab90 "auxiliary test code (environments or event listeners)") at /home/jdillaman/ceph_wip/src/googletest/googletest/src/gtest.cc:2438
#28 0x00005555564977ad in testing::UnitTest::Run() () at /home/jdillaman/ceph_wip/src/googletest/googletest/include/gtest/gtest.h:1287
#29 0x0000555555b736be in RUN_ALL_TESTS () at /home/jdillaman/ceph_wip/src/googletest/googletest/include/gtest/gtest.h:2233
#30 main () at /home/jdillaman/ceph_wip/src/test/librbd/test_main.cc:74
#31 0x00007fffeb21e11b in __libc_start_main () from /lib64/libc.so.6
#32 0x0000555555b9fbca in _start () at /usr/include/c++/8/bits/char_traits.h:285
```

MahatiC commented:
@dillaman could you let me know the command you had used? I tried running it gdb:

Reading symbols from ./bin/unittest_librbd...done.
(gdb) set args --gtest_filter=TestMockDeepCopyRequest.SimpleCopy
(gdb) run
Starting program: /home/otc/ceph/build/bin/unittest_librbd --gtest_filter=TestMockDeepCopyRequest.SimpleCopy
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe95ba700 (LWP 6591)]
[New Thread 0x7fffe8db9700 (LWP 6592)]
[New Thread 0x7fffe85b8700 (LWP 6593)]
[New Thread 0x7fffe7db7700 (LWP 6594)]
[New Thread 0x7fffe75b6700 (LWP 6595)]
[New Thread 0x7fffe6db5700 (LWP 6596)]
[New Thread 0x7fffe65b4700 (LWP 6597)]
[New Thread 0x7fffe5db3700 (LWP 6598)]
[New Thread 0x7fffe55b2700 (LWP 6599)]
[New Thread 0x7fffe4db1700 (LWP 6600)]
[New Thread 0x7fffe45b0700 (LWP 6601)]
[New Thread 0x7fffe3daf700 (LWP 6602)]
[New Thread 0x7fffe35ae700 (LWP 6603)]
[New Thread 0x7fffe2dad700 (LWP 6604)]
[New Thread 0x7fffe25ac700 (LWP 6605)]
[New Thread 0x7fffe1dab700 (LWP 6606)]
[New Thread 0x7fffe15aa700 (LWP 6607)]
[New Thread 0x7fffe0da9700 (LWP 6608)]
[New Thread 0x7fffe05a8700 (LWP 6609)]
[New Thread 0x7fffdfda7700 (LWP 6610)]
[New Thread 0x7fffdf5a6700 (LWP 6611)]
[New Thread 0x7fffdeda5700 (LWP 6612)]
[New Thread 0x7fffde5a4700 (LWP 6613)]
[New Thread 0x7fffddda3700 (LWP 6614)]
[New Thread 0x7fffdd5a2700 (LWP 6615)]
[New Thread 0x7fffdcda1700 (LWP 6616)]
[New Thread 0x7fffdc5a0700 (LWP 6617)]
[New Thread 0x7fffdbd9f700 (LWP 6618)]
[New Thread 0x7fffdb59e700 (LWP 6619)]
[New Thread 0x7fffdad9d700 (LWP 6620)]
[New Thread 0x7fffda59c700 (LWP 6621)]
[New Thread 0x7fffd9d9b700 (LWP 6622)]
[New Thread 0x7fffd959a700 (LWP 6623)]
[New Thread 0x7fffd8d99700 (LWP 6624)]
[New Thread 0x7fffd8598700 (LWP 6625)]
[New Thread 0x7fffd7d97700 (LWP 6626)]
[New Thread 0x7fffd7596700 (LWP 6627)]
[New Thread 0x7fffd6d95700 (LWP 6628)]
[New Thread 0x7fffd6594700 (LWP 6629)]
[New Thread 0x7fffd5d93700 (LWP 6630)]
[New Thread 0x7fffd5592700 (LWP 6631)]
[New Thread 0x7fffd4d91700 (LWP 6632)]
[New Thread 0x7fffd4590700 (LWP 6633)]
[New Thread 0x7fffd3d8f700 (LWP 6634)]
[New Thread 0x7fffd358e700 (LWP 6635)]
[New Thread 0x7fffd2d8d700 (LWP 6636)]
[New Thread 0x7fffd258c700 (LWP 6637)]
[New Thread 0x7fffd1d8b700 (LWP 6638)]
[New Thread 0x7fffd158a700 (LWP 6639)]
[New Thread 0x7fffd0d89700 (LWP 6640)]
[New Thread 0x7fffd0588700 (LWP 6641)]
[New Thread 0x7fffcfd87700 (LWP 6642)]
[New Thread 0x7fffcf586700 (LWP 6643)]
[New Thread 0x7fffced85700 (LWP 6644)]
[New Thread 0x7fffce584700 (LWP 6645)]
[New Thread 0x7fffcdd83700 (LWP 6646)]
[New Thread 0x7fffcd582700 (LWP 6647)]
[New Thread 0x7fffccd81700 (LWP 6648)]
[New Thread 0x7fffcc580700 (LWP 6649)]
[New Thread 0x7fffcbd7f700 (LWP 6650)]
[New Thread 0x7fffcb57e700 (LWP 6651)]
[New Thread 0x7fffcad7d700 (LWP 6652)]
[New Thread 0x7fffca57c700 (LWP 6653)]
[New Thread 0x7fffc9d7b700 (LWP 6654)]
[New Thread 0x7fffc957a700 (LWP 6655)]
[New Thread 0x7fffc8d79700 (LWP 6656)]
[New Thread 0x7fffc8578700 (LWP 6657)]
[New Thread 0x7fffc7d77700 (LWP 6658)]
[New Thread 0x7fffc7576700 (LWP 6659)]
[New Thread 0x7fffc6d75700 (LWP 6660)]
[New Thread 0x7fffc6574700 (LWP 6661)]
[New Thread 0x7fffc5d73700 (LWP 6662)]
[New Thread 0x7fffc5572700 (LWP 6663)]
[New Thread 0x7fffc4d71700 (LWP 6664)]
[New Thread 0x7fffc4570700 (LWP 6665)]
[New Thread 0x7fffc3d6f700 (LWP 6666)]
[New Thread 0x7fffc356e700 (LWP 6667)]
[New Thread 0x7fffc2d6d700 (LWP 6668)]
[New Thread 0x7fffc256c700 (LWP 6669)]
[New Thread 0x7fffc1d6b700 (LWP 6670)]
[New Thread 0x7fffc156a700 (LWP 6671)]
[New Thread 0x7fffc0d69700 (LWP 6672)]
[New Thread 0x7fffc0568700 (LWP 6673)]
[New Thread 0x7fffbfd67700 (LWP 6674)]
[New Thread 0x7fffbf566700 (LWP 6675)]
[New Thread 0x7fffbed65700 (LWP 6676)]
[New Thread 0x7fffbe564700 (LWP 6677)]
[New Thread 0x7fffbdd63700 (LWP 6678)]
[New Thread 0x7fffbd562700 (LWP 6679)]
[New Thread 0x7fffbcd61700 (LWP 6680)]
seed 6587
Note: Google Test filter = TestMockDeepCopyRequest.SimpleCopy
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TestMockDeepCopyRequest
[New Thread 0x7fffbbaff700 (LWP 6681)]
[New Thread 0x7fffbb2fe700 (LWP 6682)]
[New Thread 0x7fffbaafd700 (LWP 6683)]
[New Thread 0x7fffba2fc700 (LWP 6684)]
[New Thread 0x7fffb9afb700 (LWP 6685)]
[New Thread 0x7fffb92fa700 (LWP 6686)]
[New Thread 0x7fffb8af9700 (LWP 6687)]
[New Thread 0x7fffb82f8700 (LWP 6688)]
[New Thread 0x7fffb7af7700 (LWP 6689)]
[New Thread 0x7fffb72f6700 (LWP 6690)]
[New Thread 0x7fffb6af5700 (LWP 6691)]
[New Thread 0x7fffb62f4700 (LWP 6692)]
[New Thread 0x7fffb5af3700 (LWP 6693)]
[New Thread 0x7fffb52f2700 (LWP 6694)]
[New Thread 0x7fffb4af1700 (LWP 6695)]
[New Thread 0x7fffb42f0700 (LWP 6696)]
[New Thread 0x7fffb3aef700 (LWP 6697)]
[New Thread 0x7fffb32ee700 (LWP 6698)]
[New Thread 0x7fffb2aed700 (LWP 6699)]
[New Thread 0x7fffb22ec700 (LWP 6700)]
[New Thread 0x7fffb1aeb700 (LWP 6701)]
[New Thread 0x7fffb12ea700 (LWP 6702)]
[New Thread 0x7fffb0ae9700 (LWP 6703)]
[New Thread 0x7fffb02e8700 (LWP 6704)]
[New Thread 0x7fffafae7700 (LWP 6705)]
[New Thread 0x7fffaf2e6700 (LWP 6706)]
[New Thread 0x7fffaeae5700 (LWP 6707)]
[New Thread 0x7fffae2e4700 (LWP 6708)]
[New Thread 0x7fffadae3700 (LWP 6709)]
[New Thread 0x7fffad2e2700 (LWP 6710)]
[New Thread 0x7fffacae1700 (LWP 6711)]
[New Thread 0x7fffac2e0700 (LWP 6712)]
[New Thread 0x7fffabadf700 (LWP 6713)]
[New Thread 0x7fffab2de700 (LWP 6714)]
[New Thread 0x7fffaaadd700 (LWP 6715)]
[New Thread 0x7fffaa2dc700 (LWP 6716)]
[New Thread 0x7fffa9adb700 (LWP 6717)]
[New Thread 0x7fffa92da700 (LWP 6718)]
[New Thread 0x7fffa8ad9700 (LWP 6719)]
[New Thread 0x7fffa82d8700 (LWP 6720)]
[New Thread 0x7fffa7ad7700 (LWP 6721)]
[New Thread 0x7fffa72d6700 (LWP 6722)]
[New Thread 0x7fffa6ad5700 (LWP 6723)]
[New Thread 0x7fffa62d4700 (LWP 6724)]
[New Thread 0x7fffa5ad3700 (LWP 6725)]
[New Thread 0x7fffa52d2700 (LWP 6726)]
[New Thread 0x7fffa4ad1700 (LWP 6727)]
[New Thread 0x7fffa42d0700 (LWP 6728)]
[New Thread 0x7fffa3acf700 (LWP 6729)]
[New Thread 0x7fffa32ce700 (LWP 6730)]
[New Thread 0x7fffa2acd700 (LWP 6731)]
[New Thread 0x7fffa22cc700 (LWP 6732)]
[New Thread 0x7fffa1acb700 (LWP 6733)]
[New Thread 0x7fffa12ca700 (LWP 6734)]
[New Thread 0x7fffa0ac9700 (LWP 6735)]
[New Thread 0x7fffa02c8700 (LWP 6736)]
[New Thread 0x7fff9fac7700 (LWP 6737)]
[New Thread 0x7fff9f2c6700 (LWP 6738)]
[New Thread 0x7fff9eac5700 (LWP 6739)]
[New Thread 0x7fff9e2c4700 (LWP 6740)]
[New Thread 0x7fff9dac3700 (LWP 6741)]
[New Thread 0x7fff9d2c2700 (LWP 6742)]
[New Thread 0x7fff9cac1700 (LWP 6743)]
[New Thread 0x7fff9c2c0700 (LWP 6744)]
[New Thread 0x7fff9babf700 (LWP 6745)]
[New Thread 0x7fff9b2be700 (LWP 6746)]
[New Thread 0x7fff9aabd700 (LWP 6747)]
[New Thread 0x7fff9a2bc700 (LWP 6748)]
[New Thread 0x7fff99abb700 (LWP 6749)]
[New Thread 0x7fff992ba700 (LWP 6750)]
[New Thread 0x7fff98ab9700 (LWP 6751)]
[New Thread 0x7fff982b8700 (LWP 6752)]
[New Thread 0x7fff97ab7700 (LWP 6753)]
[New Thread 0x7fff972b6700 (LWP 6754)]
[New Thread 0x7fff96ab5700 (LWP 6755)]
[New Thread 0x7fff962b4700 (LWP 6756)]
[New Thread 0x7fff95ab3700 (LWP 6757)]
[New Thread 0x7fff952b2700 (LWP 6758)]
[New Thread 0x7fff94ab1700 (LWP 6759)]
[New Thread 0x7fff942b0700 (LWP 6760)]
[New Thread 0x7fff93aaf700 (LWP 6761)]
[New Thread 0x7fff932ae700 (LWP 6762)]
[New Thread 0x7fff92aad700 (LWP 6763)]
[New Thread 0x7fff922ac700 (LWP 6764)]
[New Thread 0x7fff91aab700 (LWP 6765)]
[New Thread 0x7fff912aa700 (LWP 6766)]
[New Thread 0x7fff90aa9700 (LWP 6767)]
[New Thread 0x7fff902a8700 (LWP 6768)]
[New Thread 0x7fff8faa7700 (LWP 6769)]
[New Thread 0x7fff8f2a6700 (LWP 6770)]
[New Thread 0x7fff8e044700 (LWP 6771)]
[New Thread 0x7fff8d843700 (LWP 6772)]
[New Thread 0x7fff8d042700 (LWP 6773)]
[New Thread 0x7fff8c841700 (LWP 6774)]
[New Thread 0x7fff8c040700 (LWP 6775)]
[New Thread 0x7fff8b83f700 (LWP 6776)]
[New Thread 0x7fff8b03e700 (LWP 6777)]
[New Thread 0x7fff8a83d700 (LWP 6778)]
[New Thread 0x7fff8a03c700 (LWP 6779)]
[New Thread 0x7fff8983b700 (LWP 6780)]
[New Thread 0x7fff8903a700 (LWP 6781)]
[New Thread 0x7fff88839700 (LWP 6782)]
[New Thread 0x7fff88038700 (LWP 6783)]
[New Thread 0x7fff87837700 (LWP 6784)]
[New Thread 0x7fff87036700 (LWP 6785)]
[New Thread 0x7fff86835700 (LWP 6786)]
[New Thread 0x7fff86034700 (LWP 6787)]
[New Thread 0x7fff85833700 (LWP 6788)]
[New Thread 0x7fff85032700 (LWP 6789)]
[New Thread 0x7fff84831700 (LWP 6790)]
[New Thread 0x7fff84030700 (LWP 6791)]
[New Thread 0x7fff8382f700 (LWP 6792)]
[New Thread 0x7fff8302e700 (LWP 6793)]
[New Thread 0x7fff8282d700 (LWP 6794)]
[New Thread 0x7fff8202c700 (LWP 6795)]
[New Thread 0x7fff8182b700 (LWP 6796)]
[New Thread 0x7fff8102a700 (LWP 6797)]
[New Thread 0x7fff80829700 (LWP 6798)]
[New Thread 0x7fff80028700 (LWP 6799)]
[New Thread 0x7fff7f827700 (LWP 6800)]
[New Thread 0x7fff7f026700 (LWP 6801)]
[New Thread 0x7fff7e825700 (LWP 6802)]
[New Thread 0x7fff7e024700 (LWP 6803)]
[New Thread 0x7fff7d823700 (LWP 6804)]
[New Thread 0x7fff7d022700 (LWP 6805)]
[New Thread 0x7fff7c821700 (LWP 6806)]
[New Thread 0x7fff7c020700 (LWP 6807)]
[New Thread 0x7fff7b81f700 (LWP 6808)]
[New Thread 0x7fff7b01e700 (LWP 6809)]
[New Thread 0x7fff7a81d700 (LWP 6810)]
[New Thread 0x7fff7a01c700 (LWP 6811)]
[New Thread 0x7fff7981b700 (LWP 6812)]
[New Thread 0x7fff7901a700 (LWP 6813)]
[New Thread 0x7fff78819700 (LWP 6814)]
[New Thread 0x7fff78018700 (LWP 6815)]
[New Thread 0x7fff77817700 (LWP 6816)]
[New Thread 0x7fff77016700 (LWP 6817)]
[New Thread 0x7fff76815700 (LWP 6818)]
[New Thread 0x7fff76014700 (LWP 6819)]
[New Thread 0x7fff75813700 (LWP 6820)]
[New Thread 0x7fff75012700 (LWP 6821)]
[New Thread 0x7fff74811700 (LWP 6822)]
[New Thread 0x7fff74010700 (LWP 6823)]
[New Thread 0x7fff7380f700 (LWP 6824)]
[New Thread 0x7fff7300e700 (LWP 6825)]
[New Thread 0x7fff7280d700 (LWP 6826)]
[New Thread 0x7fff7200c700 (LWP 6827)]
[New Thread 0x7fff7180b700 (LWP 6828)]
[New Thread 0x7fff7100a700 (LWP 6829)]
[New Thread 0x7fff70809700 (LWP 6830)]
[New Thread 0x7fff70008700 (LWP 6831)]
[New Thread 0x7fff6f807700 (LWP 6832)]
[New Thread 0x7fff6f006700 (LWP 6833)]
[New Thread 0x7fff6e805700 (LWP 6834)]
[New Thread 0x7fff6e004700 (LWP 6835)]
[New Thread 0x7fff6d803700 (LWP 6836)]
[New Thread 0x7fff6d002700 (LWP 6837)]
[New Thread 0x7fff6c801700 (LWP 6838)]
[New Thread 0x7fff6c000700 (LWP 6839)]
[New Thread 0x7fff6b7ff700 (LWP 6840)]
[New Thread 0x7fff6affe700 (LWP 6841)]
[New Thread 0x7fff6a7fd700 (LWP 6842)]
[New Thread 0x7fff69ffc700 (LWP 6843)]
[New Thread 0x7fff697fb700 (LWP 6844)]
[New Thread 0x7fff68ffa700 (LWP 6845)]
[New Thread 0x7fff687f9700 (LWP 6846)]
[New Thread 0x7fff67ff8700 (LWP 6847)]
[New Thread 0x7fff677f7700 (LWP 6848)]
[New Thread 0x7fff66ff6700 (LWP 6849)]
[New Thread 0x7fff667f5700 (LWP 6850)]
[New Thread 0x7fff65ff4700 (LWP 6851)]
[New Thread 0x7fff657f3700 (LWP 6852)]
[New Thread 0x7fff64ff2700 (LWP 6853)]
[New Thread 0x7fff647f1700 (LWP 6854)]
[New Thread 0x7fff63ff0700 (LWP 6855)]
[New Thread 0x7fff637ef700 (LWP 6856)]
[New Thread 0x7fff62fee700 (LWP 6857)]
[New Thread 0x7fff627ed700 (LWP 6858)]
[New Thread 0x7fff61fec700 (LWP 6859)]
[New Thread 0x7fff617eb700 (LWP 6860)]
[ RUN      ] TestMockDeepCopyRequest.SimpleCopy
[New Thread 0x7fff60fea700 (LWP 6861)]
[New Thread 0x7fff5bfff700 (LWP 6862)]
[New Thread 0x7fff5b7fe700 (LWP 6863)]
[New Thread 0x7fff5affd700 (LWP 6864)]
Failed to load class: cas (lib/libcls_cas.so): lib/libcls_cas.so: undefined symbol: _Z14cls_cxx_removePv
Failed to load class: cephfs (lib/libcls_cephfs.so): lib/libcls_cephfs.so: undefined symbol: _ZTI10PGLSFilter
Failed to load class: hello (lib/libcls_hello.so): lib/libcls_hello.so: undefined symbol: _ZTI10PGLSFilter
Failed to load class: log (lib/libcls_log.so): lib/libcls_log.so: undefined symbol: _Z23cls_cxx_map_read_headerPvPN4ceph6buffer4listE
Failed to load class: lua (lib/libcls_lua.so): lib/libcls_lua.so: undefined symbol: _Z24cls_cxx_map_write_headerPvPN4ceph6buffer4listE
Failed to load class: refcount (lib/libcls_refcount.so): lib/libcls_refcount.so: undefined symbol: _Z14cls_cxx_removePv
Failed to load class: rgw (lib/libcls_rgw.so): lib/libcls_rgw.so: undefined symbol: _Z24cls_cxx_map_write_headerPvPN4ceph6buffer4listE
Failed to load class: sdk (lib/libcls_sdk.so): lib/libcls_sdk.so: undefined symbol: _Z14cls_cxx_removePv
Failed to load class: user (lib/libcls_user.so): lib/libcls_user.so: undefined symbol: _Z23cls_cxx_map_read_headerPvPN4ceph6buffer4listE
Failed to load class: version (lib/libcls_version.so): lib/libcls_version.so: undefined symbol: _Z19cls_gen_rand_base64Pci
[New Thread 0x7fff59cb8700 (LWP 6865)]
[New Thread 0x7fff594b7700 (LWP 6866)]
[New Thread 0x7fff58cb6700 (LWP 6867)]
[New Thread 0x7fff47fff700 (LWP 6868)]
[Thread 0x7fff594b7700 (LWP 6866) exited]
[Thread 0x7fff59cb8700 (LWP 6865) exited]
[Thread 0x7fff47fff700 (LWP 6868) exited]
[Thread 0x7fff58cb6700 (LWP 6867) exited]
[       OK ] TestMockDeepCopyRequest.SimpleCopy (433 ms)
[Thread 0x7fffbb2fe700 (LWP 6682) exited]
[Thread 0x7fffba2fc700 (LWP 6684) exited]
[Thread 0x7fffb9afb700 (LWP 6685) exited]
[Thread 0x7fffb92fa700 (LWP 6686) exited]
[Thread 0x7fffbaafd700 (LWP 6683) exited]
[Thread 0x7fffb82f8700 (LWP 6688) exited]
[Thread 0x7fffb7af7700 (LWP 6689) exited]
[Thread 0x7fffb8af9700 (LWP 6687) exited]
[Thread 0x7fffb72f6700 (LWP 6690) exited]
[Thread 0x7fffb6af5700 (LWP 6691) exited]
[Thread 0x7fffb62f4700 (LWP 6692) exited]
[Thread 0x7fffb52f2700 (LWP 6694) exited]
[Thread 0x7fffb4af1700 (LWP 6695) exited]
[Thread 0x7fffb5af3700 (LWP 6693) exited]
[Thread 0x7fffb32ee700 (LWP 6698) exited]
[Thread 0x7fffb3aef700 (LWP 6697) exited]
[Thread 0x7fffb2aed700 (LWP 6699) exited]
[Thread 0x7fffb42f0700 (LWP 6696) exited]
[Thread 0x7fffb22ec700 (LWP 6700) exited]
[Thread 0x7fffb12ea700 (LWP 6702) exited]
[Thread 0x7fffb1aeb700 (LWP 6701) exited]
[Thread 0x7fffb0ae9700 (LWP 6703) exited]
[Thread 0x7fffafae7700 (LWP 6705) exited]
[Thread 0x7fffb02e8700 (LWP 6704) exited]
[Thread 0x7fffaf2e6700 (LWP 6706) exited]
[Thread 0x7fffae2e4700 (LWP 6708) exited]
[Thread 0x7fffadae3700 (LWP 6709) exited]
[Thread 0x7fffad2e2700 (LWP 6710) exited]
[Thread 0x7fffaeae5700 (LWP 6707) exited]
[Thread 0x7fffac2e0700 (LWP 6712) exited]
[Thread 0x7fffacae1700 (LWP 6711) exited]
[Thread 0x7fffabadf700 (LWP 6713) exited]
[Thread 0x7fffaaadd700 (LWP 6715) exited]
[Thread 0x7fffab2de700 (LWP 6714) exited]
[Thread 0x7fffaa2dc700 (LWP 6716) exited]
[Thread 0x7fffa9adb700 (LWP 6717) exited]
[Thread 0x7fffa8ad9700 (LWP 6719) exited]
[Thread 0x7fffa82d8700 (LWP 6720) exited]
[Thread 0x7fffa92da700 (LWP 6718) exited]
[Thread 0x7fffa72d6700 (LWP 6722) exited]
[Thread 0x7fffa7ad7700 (LWP 6721) exited]
[Thread 0x7fffa6ad5700 (LWP 6723) exited]
[Thread 0x7fffa62d4700 (LWP 6724) exited]
[Thread 0x7fffa5ad3700 (LWP 6725) exited]
[Thread 0x7fffa52d2700 (LWP 6726) exited]
[Thread 0x7fffa4ad1700 (LWP 6727) exited]
[Thread 0x7fffa42d0700 (LWP 6728) exited]
[Thread 0x7fffa3acf700 (LWP 6729) exited]
[Thread 0x7fffa32ce700 (LWP 6730) exited]
[Thread 0x7fffa2acd700 (LWP 6731) exited]
[Thread 0x7fffa22cc700 (LWP 6732) exited]
[Thread 0x7fffa1acb700 (LWP 6733) exited]
[Thread 0x7fffa0ac9700 (LWP 6735) exited]
[Thread 0x7fffa12ca700 (LWP 6734) exited]
[Thread 0x7fffa02c8700 (LWP 6736) exited]
[Thread 0x7fff9f2c6700 (LWP 6738) exited]
[Thread 0x7fff9eac5700 (LWP 6739) exited]
[Thread 0x7fff9fac7700 (LWP 6737) exited]
[Thread 0x7fff9e2c4700 (LWP 6740) exited]
[Thread 0x7fff9dac3700 (LWP 6741) exited]
[Thread 0x7fff9d2c2700 (LWP 6742) exited]
[Thread 0x7fff9cac1700 (LWP 6743) exited]
[Thread 0x7fff9c2c0700 (LWP 6744) exited]
[Thread 0x7fff9babf700 (LWP 6745) exited]
[Thread 0x7fff9b2be700 (LWP 6746) exited]
[Thread 0x7fff9aabd700 (LWP 6747) exited]
[Thread 0x7fff9a2bc700 (LWP 6748) exited]
[Thread 0x7fff99abb700 (LWP 6749) exited]
[Thread 0x7fff992ba700 (LWP 6750) exited]
[Thread 0x7fff98ab9700 (LWP 6751) exited]
[Thread 0x7fff982b8700 (LWP 6752) exited]
[Thread 0x7fff97ab7700 (LWP 6753) exited]
[Thread 0x7fff972b6700 (LWP 6754) exited]
[Thread 0x7fff96ab5700 (LWP 6755) exited]
[Thread 0x7fff962b4700 (LWP 6756) exited]
[Thread 0x7fff95ab3700 (LWP 6757) exited]
[Thread 0x7fff952b2700 (LWP 6758) exited]
[Thread 0x7fff94ab1700 (LWP 6759) exited]
[Thread 0x7fff942b0700 (LWP 6760) exited]
[Thread 0x7fff93aaf700 (LWP 6761) exited]
[Thread 0x7fff932ae700 (LWP 6762) exited]
[Thread 0x7fff91aab700 (LWP 6765) exited]
[Thread 0x7fff922ac700 (LWP 6764) exited]
[Thread 0x7fff92aad700 (LWP 6763) exited]
[Thread 0x7fff912aa700 (LWP 6766) exited]
[Thread 0x7fff90aa9700 (LWP 6767) exited]
[Thread 0x7fff902a8700 (LWP 6768) exited]
[Thread 0x7fff8f2a6700 (LWP 6770) exited]
[Thread 0x7fff8faa7700 (LWP 6769) exited]
[Thread 0x7fffbbaff700 (LWP 6681) exited]
[Thread 0x7fff8d843700 (LWP 6772) exited]
[Thread 0x7fff8d042700 (LWP 6773) exited]
[Thread 0x7fff8c040700 (LWP 6775) exited]
[Thread 0x7fff8b83f700 (LWP 6776) exited]
[Thread 0x7fff8b03e700 (LWP 6777) exited]
[Thread 0x7fff8c841700 (LWP 6774) exited]
[Thread 0x7fff8a83d700 (LWP 6778) exited]
[Thread 0x7fff8a03c700 (LWP 6779) exited]
[Thread 0x7fff8983b700 (LWP 6780) exited]
[Thread 0x7fff8903a700 (LWP 6781) exited]
[Thread 0x7fff88839700 (LWP 6782) exited]
[Thread 0x7fff88038700 (LWP 6783) exited]
[Thread 0x7fff87837700 (LWP 6784) exited]
[Thread 0x7fff87036700 (LWP 6785) exited]
[Thread 0x7fff86835700 (LWP 6786) exited]
[Thread 0x7fff86034700 (LWP 6787) exited]
[Thread 0x7fff85833700 (LWP 6788) exited]
[Thread 0x7fff85032700 (LWP 6789) exited]
[Thread 0x7fff84831700 (LWP 6790) exited]
[Thread 0x7fff84030700 (LWP 6791) exited]
[Thread 0x7fff8302e700 (LWP 6793) exited]
[Thread 0x7fff8282d700 (LWP 6794) exited]
[Thread 0x7fff8382f700 (LWP 6792) exited]
[Thread 0x7fff8182b700 (LWP 6796) exited]
[Thread 0x7fff8202c700 (LWP 6795) exited]
[Thread 0x7fff8102a700 (LWP 6797) exited]
[Thread 0x7fff80829700 (LWP 6798) exited]
[Thread 0x7fff80028700 (LWP 6799) exited]
[Thread 0x7fff7f026700 (LWP 6801) exited]
[Thread 0x7fff7f827700 (LWP 6800) exited]
[Thread 0x7fff7e825700 (LWP 6802) exited]
[Thread 0x7fff7e024700 (LWP 6803) exited]
[Thread 0x7fff7d823700 (LWP 6804) exited]
[Thread 0x7fff7d022700 (LWP 6805) exited]
[Thread 0x7fff7c821700 (LWP 6806) exited]
[Thread 0x7fff7c020700 (LWP 6807) exited]
[Thread 0x7fff7b81f700 (LWP 6808) exited]
[Thread 0x7fff7b01e700 (LWP 6809) exited]
[Thread 0x7fff7a81d700 (LWP 6810) exited]
[Thread 0x7fff7a01c700 (LWP 6811) exited]
[Thread 0x7fff7981b700 (LWP 6812) exited]
[Thread 0x7fff7901a700 (LWP 6813) exited]
[Thread 0x7fff78819700 (LWP 6814) exited]
[Thread 0x7fff78018700 (LWP 6815) exited]
[Thread 0x7fff77817700 (LWP 6816) exited]
[Thread 0x7fff77016700 (LWP 6817) exited]
[Thread 0x7fff76815700 (LWP 6818) exited]
[Thread 0x7fff76014700 (LWP 6819) exited]
[Thread 0x7fff75813700 (LWP 6820) exited]
[Thread 0x7fff75012700 (LWP 6821) exited]
[Thread 0x7fff74811700 (LWP 6822) exited]
[Thread 0x7fff74010700 (LWP 6823) exited]
[Thread 0x7fff7380f700 (LWP 6824) exited]
[Thread 0x7fff7300e700 (LWP 6825) exited]
[Thread 0x7fff7280d700 (LWP 6826) exited]
[Thread 0x7fff7200c700 (LWP 6827) exited]
[Thread 0x7fff7180b700 (LWP 6828) exited]
[Thread 0x7fff7100a700 (LWP 6829) exited]
[Thread 0x7fff70809700 (LWP 6830) exited]
[Thread 0x7fff70008700 (LWP 6831) exited]
[Thread 0x7fff6f807700 (LWP 6832) exited]
[Thread 0x7fff6f006700 (LWP 6833) exited]
[Thread 0x7fff6e805700 (LWP 6834) exited]
[Thread 0x7fff6e004700 (LWP 6835) exited]
[Thread 0x7fff6d803700 (LWP 6836) exited]
[Thread 0x7fff6d002700 (LWP 6837) exited]
[Thread 0x7fff6c801700 (LWP 6838) exited]
[Thread 0x7fff6c000700 (LWP 6839) exited]
[Thread 0x7fff6b7ff700 (LWP 6840) exited]
[Thread 0x7fff6affe700 (LWP 6841) exited]
[Thread 0x7fff6a7fd700 (LWP 6842) exited]
[Thread 0x7fff69ffc700 (LWP 6843) exited]
[Thread 0x7fff697fb700 (LWP 6844) exited]
[Thread 0x7fff68ffa700 (LWP 6845) exited]
[Thread 0x7fff687f9700 (LWP 6846) exited]
[Thread 0x7fff67ff8700 (LWP 6847) exited]
[Thread 0x7fff677f7700 (LWP 6848) exited]
[Thread 0x7fff66ff6700 (LWP 6849) exited]
[Thread 0x7fff667f5700 (LWP 6850) exited]
[Thread 0x7fff65ff4700 (LWP 6851) exited]
[Thread 0x7fff657f3700 (LWP 6852) exited]
[Thread 0x7fff64ff2700 (LWP 6853) exited]
[Thread 0x7fff647f1700 (LWP 6854) exited]
[Thread 0x7fff63ff0700 (LWP 6855) exited]
[Thread 0x7fff637ef700 (LWP 6856) exited]
[Thread 0x7fff62fee700 (LWP 6857) exited]
[Thread 0x7fff627ed700 (LWP 6858) exited]
[Thread 0x7fff61fec700 (LWP 6859) exited]
[Thread 0x7fff617eb700 (LWP 6860) exited]
[Thread 0x7fff60fea700 (LWP 6861) exited]
[Thread 0x7fff5bfff700 (LWP 6862) exited]
[Thread 0x7fff5b7fe700 (LWP 6863) exited]
[Thread 0x7fff5affd700 (LWP 6864) exited]
[Thread 0x7fff8e044700 (LWP 6771) exited]
[----------] 1 test from TestMockDeepCopyRequest (433 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (562 ms total)
[  PASSED  ] 1 test.

dillaman commented:
@MahatiC the ```ThreadPoolSingleton``` was throwing an immediate exception since a uint64_t field was being accessed as an int64_t field.

```
diff --git a/src/librbd/ImageCtx.cc b/src/librbd/ImageCtx.cc
index 308502e7cd..a9f1af999a 100644
--- a/src/librbd/ImageCtx.cc
+++ b/src/librbd/ImageCtx.cc
@@ -61,7 +61,7 @@ public:
 
   explicit ThreadPoolSingleton(CephContext *cct)
     : ThreadPool(cct, "librbd::thread_pool", "tp_librbd",
-                 cct->_conf.get_val<int64_t>("rbd_op_threads"), "rbd_op_threads"),
+                 cct->_conf.get_val<uint64_t>("rbd_op_threads"), "rbd_op_threads"),
       op_work_queue(new ContextWQ("librbd::op_work_queue",
                                   cct->_conf.get_val<uint64_t>("rbd_op_thread_timeout"),
                                   this)) {

```

liyichao commented on src/librbd/ExclusiveLock.cc:
```diff
@@ -101,6 +103,10 @@ int ExclusiveLock<I>::get_unlocked_op_error() const {
   ceph_assert(m_image_ctx.owner_lock.is_locked());
+
+  using klass = ExclusiveLock<I>;
+  on_init = create_ref_counted_context<klass>(this, on_init);
```
I do not quite understand this change, is this for avoiding exlusiveLock to be deleted before on_init is called? If so, in what situation will this happen?

liyichao replied:
What is the race condition? Could you give an example ?

MahatiC commented:
> @MahatiC I didn't do anything special -- it just crashes for me every time. I run Fedora 28 and I have 12 CPU cores available.

Thanks for catching that exception. With your rebase, still can't reproduce the build error. I'm going to setup the environment with 12 CPU cores and give it a go.

MahatiC commented:
@dillaman I ran these tests with 12 cores (ubuntu env itself) and can't reproduce the errors. I had one of my colleagues test out this patch as well and they reported no errors.

dillaman commented:
retest this please

MahatiC commented:
retest this please

MahatiC commented:
retest this please

MahatiC commented:
@dillaman `make check (arm64)` and `make check` succeeded before rebase (and a small conflict resolution) to master. But now `make check(arm64)` is failing. Idk if the rebase is causing this failure. Here is the diff of conflict resolution during rebase:

`diff --git a/qa/suites/rbd/thrash/workloads/rbd_fsx_rate_limit.yaml b/qa/suites/rbd/thrash/workloads/rbd_fsx_rate_limit.yaml
index 611320b..782a2af 100644
--- a/qa/suites/rbd/thrash/workloads/rbd_fsx_rate_limit.yaml
+++ b/qa/suites/rbd/thrash/workloads/rbd_fsx_rate_limit.yaml
@@ -7,5 +7,6 @@ overrides:
     conf:
       client:
         rbd qos iops limit: 50
+        rbd op threads: 8
         rbd qos iops burst: 100
         rbd qos schedule tick min: 100`

tchaikov commented:
retest this please.

MahatiC commented:
@dillaman ping, please review

dillaman commented:
@MahatiC I was out-of-office last week. I have a build running right now so that it can be run through teuthology.

dillaman commented:
http://pulpito.ceph.com/jdillaman-2018-12-03_11:08:01-rbd-wip-jd-testing-distro-basic-smithi/

dillaman commented:
@MahatiC Looks like at least one or more memory leaks [1][2] and a import/export comparison failure [3] and a crash in librbd in xlist handling [4]

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2018-12-03_11:08:01-rbd-wip-jd-testing-distro-basic-smithi/3303923/teuthology.log
[2] http://qa-proxy.ceph.com/teuthology/jdillaman-2018-12-03_11:08:01-rbd-wip-jd-testing-distro-basic-smithi/3303970/teuthology.log
[3] http://qa-proxy.ceph.com/teuthology/jdillaman-2018-12-03_11:08:01-rbd-wip-jd-testing-distro-basic-smithi/3303957/teuthology.log
[4] http://qa-proxy.ceph.com/teuthology/jdillaman-2018-12-03_11:08:01-rbd-wip-jd-testing-distro-basic-smithi/3303925/teuthology.log

MahatiC commented:
@dillaman Ah okay. Thanks much. Will look into those.

liyichao commented:
Seen from file history, the multiple threads option is disabled due to http://tracker.ceph.com/issues/15034 , not sure it is fixed.

liyichao commented:
Tracker ticket is http://tracker.ceph.com/issues/17379 for those who are interested.

MahatiC commented:
@dillaman I identified some unnecessary null pointer assignments ahead of object deletion, this is still WIP. Will push the change once I check the tests as well, fyi

MahatiC commented:
@dillaman hi, I pushed a newer version a couple of weeks back. I removed what I found as premature pointer deletions. Could you take look? Thanks.

dillaman commented:
@MahatiC One thing that I recently hit when fixing a new bug in "rbd export" was that even if librbd can internally support extra worker threads, we need to ensure that AIO callbacks are still serialized. Perhaps via a new (special) work queue that doesn't pop any entries if another thread pool thread is already processing the AIO callback work queue.

dillaman commented:
@MahatiC librbd itself should be fine (once the race conditions are fixed). Any client that wants higher throughput on the completions would need to use something like the eventfd interface where it can grab a batch of completed events at the same time across multiple threads.

MahatiC commented:
Any popping seems to happen on excessive threads[0] after issuing a `join()`. And otherwise too all threads seem to join and delete [1]. If you anticipate an issue happening when rbd_op_threads > 1, it seems likely it should already happen when rbd_op_threads = 1. Did you encounter the issue with single thread?

[0] https://github.com/ceph/ceph/blob/master/src/common/WorkQueue.cc#L167
[1] https://github.com/ceph/ceph/blob/master/src/common/WorkQueue.cc#L203-L204

MahatiC commented:
@dillaman Does the above make sense? Could you give me any more details on the rbd export bug?

dillaman commented:
@MahatiC I apologize for the delay. Right now, librbd::AioCompletions are only completed under two conditions: (1) the librados Finisher thread callback from the osdc Objecter or (2) the osdc ObjectCacher -- and right now it's always method (1) if the cache is disabled or method (2) if the cache is enabled. 

What I was trying to point out is that if the goal is to eventually support N IO dispatch threads, we need to resolve the callback issue since we can really only complete to external applications via a single thread. If everything continues to be serialized by the Objecter or ObjectCacher for AIO callbacks, there isn't an issue. But eventually, we will want to remove that restriction and when we do, we will need to serialize the AIO completions.

MahatiC commented:
@dillaman Thanks for clarifying. To serialize AIO completions, we'd need multi-threaded callback execution in place and PR#16715 is WIP to enable this. Do you think this PR is blocked by PR 16715?

I believe the work can continue on this PR in fixing the race conditions. And once there's support for multiple AIO completions, there can be another PR for serializing them (retaining rbd_op_threads = 1 in code until then). What do you think?

dillaman commented:
@MahatiC Yes, that sounds good to me. Thanks.

MahatiC commented:
@dillaman Hi, could you run this through Teuthology and paste back the results please?

dillaman commented:
@MahatiC Can you rebase this and I'll re-run it?

MahatiC commented:
> @MahatiC Can you rebase this and I'll re-run it?

My apologies, I totally missed the conflicting files below. Will rebase.

MahatiC commented:
retest this please

dillaman commented:
@MahatiC It was a very "red" run: http://pulpito.ceph.com/jdillaman-2019-03-13_08:38:02-rbd-wip-jd-testing-distro-basic-smithi/

MahatiC commented:
@dillaman hi, I used the address sanitizer (https://github.com/google/sanitizers/wiki/AddressSanitizer) and saw a bunch of direct memory leaks. Fixed the code and ran it again. This time, it reported no leaks in the code. Could you run this code through Teuthology? Additionally if you could review this code, that would be great. Thanks!

dillaman commented:
@MahatiC Sure -- I'll re-run it tomorrow

dillaman commented:
http://pulpito.ceph.com/jdillaman-2019-05-16_22:08:51-rbd-wip-jd-testing-distro-basic-smithi/

MahatiC commented:
@dillaman Thanks. I no longer see the memory leaks in teuthology logs. There are a few qemu `Command crashed: 'qemu-img convert -f qcow2 -o raw..'` errors but I'm not sure if they're related to this PR.

I do see one assert failure in test_librbd.sh for "RBD_FEATURES=125" run. I tried to run the tests locally for that error but do not see it occur.

Could you verify this please?

dillaman commented:
> I do see one assert failure in test_librbd.sh for "RBD_FEATURES=125" run. I tried to run the tests locally for that error but do not see it occur.

Anything QEMU on RHEL 7 is broken right now, along w/ the dynamic_features_test.

That one looks like a real error. It's implying that a journaled IO started for extents A~B and shortly thereafter a second IO event for an overlapping extent A'~B' occurred -- but that the first IO hit the ObjectCacher after the second IO (so the ordered journal tid was out-of-order).

MahatiC commented:
retest this please

MahatiC commented:
@dillaman I've split the rest of the code from this PR into #30397 and #28952. Please review.

MahatiC commented:
@dillaman could you run this through Teuthology?

dillaman commented on qa/cephfs/overrides/fuse/default-perm/yes.yaml:
```diff
@@ -3,3 +3,5 @@ overrides:
     conf:
       client:
         fuse default permissions: true
+        rbd skip partial discard: true
```
This is a CephFS test

MahatiC replied:
reverted

dillaman commented on qa/suites/rbd/qemu/cache/writeback.yaml:
```diff
@@ -1,3 +1,9 @@
+  ceph:
+    conf:
+      client:
+        rbd default features: 125
```
Why is this being overridden in these tests (here and below)?

MahatiC replied:
my bad, forgot to cleanup. Removed.

dillaman commented:
FYI -- the master branch has been unstable all week so I need to wait for it to get fixed before running.

MahatiC commented:
Ack. Thanks for the info.

dillaman commented on qa/suites/rbd/thrash/msgr-failures/few.yaml:
```diff
@@ -4,5 +4,8 @@ overrides:
-    log-whitelist:
+      client:
+        rbd op threads: 8
+   log-whitelist:
```
The YAML parser fails here do to the messed up indentation.

dillaman commented on qa/suites/rbd/cli_v1/msgr-failures/few.yaml:
```diff
@@ -4,5 +4,7 @@ overrides:
-    log-whitelist:
+      client:
+        rbd op threads: 8
+   log-whitelist:
```
Same here

dillaman commented:
http://pulpito.ceph.com/jdillaman-2019-11-26_08:26:12-rbd-wip-jd-testing-distro-basic-smithi/

MahatiC commented:
@dillaman Thanks. I see a bunch of failures pointing to mostly one type of error. I'm looking into that.

MahatiC commented:
@dillaman It seems the overlapping IO check needs to happen before they're queued otherwise it interferes with all of the stuff that happens before dequeuing an item (`_void_dequeue`). In the last run most of the failures were related to this. I fixed this.

One other failure was related to rbd mirror. There's a race in the test `TestInstances.NotifyRemove`. Fixed this within the test. I don't think that could happen with rbd mirror code but I'm not sure. Could you verify that please?

MahatiC commented:
@dillaman did you get a chance to look through this? If not, meanwhile could you run this through Teuthology?

dillaman commented on src/test/rbd_mirror/test_Instances.cc:
```diff
@@ -131,6 +133,7 @@ TEST_F(TestInstances, NotifyRemove)
   ASSERT_EQ(0, on_init.wait());
 
   instances.acked({instance_id1, instance_id2});
+  usleep(25000);
```
Nit: this is racy -- the real cause of the issue needs to be fixed.

MahatiC replied:
I think this is related to the test: rbd_mirror code[0] calls part of the test[1] after unlocking a mutex. Also correct me if I'm wrong, `usleep` also is used in the test below in line#149 for a similar reason?

[0] https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/Instances.cc#L147-L149
[1] https://github.com/ceph/ceph/blob/master/src/test/rbd_mirror/test_Instances.cc#L49-L51

dillaman replied:
Your change only sleeps once which just seems racy -- the other case sleeps up to 100 times (25 seconds)

dillaman commented:
> @dillaman It seems the overlapping IO check needs to happen before they're queued otherwise it interferes with all of the stuff that happens before dequeuing an item (`_void_dequeue`). In the last run most of the failures were related to this. I fixed this.

What exactly was the interference? I'd really like to avoid the duplicate checks.

MahatiC commented:
> What exactly was the interference? I'd really like to avoid the duplicate checks.

There aren't duplicate checks. Instead of checking for overlapping after dequeing a req, it now checks before queuing a request.

One example of interference: when an IO is dequeuing, it acquires all the necessary locks and starts the `m_aio_comp->start_op()`. And during `process(req)` if we decide to block the IO since it overlaps and later on, when it's unblocked and queued back, it fails since `m_aio_comp->start_op()` was already started

dillaman commented on src/test/rbd_mirror/test_Instances.cc:
```diff
@@ -35,13 +35,15 @@ class TestInstances : public ::rbd::mirror::TestFixture {
-        instance->ids.insert(instance_id);
-        if (instance->count == 0) {
-          instance->ctx.complete(0);
+      if(instance->count > 0) {
```
How do you get to this point where you can invoke the callback before recording that you expect the callback?

MahatiC commented:
> it might have been due to the old oddity of how flushes where tracked

This seems to be the reason. If `start_op` is moved to `aio_XYZ` before queuing, it's causing deadlocks due to the way flushes are tracked. `TestLibRBD.FlushEmptyOpsOnExternalSnapshot` test demonstrates that.

dillaman commented:
OK, I understand what you are referring to now. What about moving the `start_op` calls to `::process_io` after it passes the blocked IO check?

MahatiC commented:
> What about moving the `start_op` calls to `::process_io` after it passes the blocked IO check?

This seems to be working ok after some tweaks. I pushed an update on this.

dillaman commented:
I'll re-run it through teuthology tonight after it builds. In the meantime, can you work to address [1]?

[1] https://github.com/ceph/ceph/pull/24483#pullrequestreview-319132083

MahatiC commented:
@dillaman Sure, thanks.

dillaman commented:
http://qa-proxy.ceph.com/teuthology/jdillaman-2020-01-24_07:38:15-rbd-wip-jd-testing-distro-basic-smithi/4700802/teuthology.log

MahatiC commented:
@dillaman hi, in `read_flags.sh`, this command is crashing on overlapping object extents when used with `WriteAround` cache --> `rbd export test@snap - --log-file test.log --debug-ms 1 --rbd-localize-snap-reads`. I suppose that's because they aren't handled in `WriteAround` cache mode. Is that correct?

It doesn't crash with in `writethrough` mode when it's using `ObjectCacher`

dillaman commented:
When I looked at it, it was crashing deep within the Objecter and the CRUSH map lookup. Those data structures *should* be protected from concurrent update and access, so it's a bit strange. There shouldn't be any restriction about sending multiple ops to the same backing object (at least crash-wise) -- they are just processed in submitted order by the OSD.

MahatiC commented:
@dillaman My traceback also always showed a crash in Objecter and crushmap. But perhaps I (wrongly) attributed it to mishandling in librbd since that was the originating thread (and due to the differing output in cache modes). But after debugging, it doesn't look like it's an issue in the cache implementation.

zhengchengyao commented:
@MahatiC @dillaman   Can this PR improve performance? I set rbd_op_threads=8，but performance has not improved，thanks

MahatiC commented:
@dillaman regarding review comment https://github.com/ceph/ceph/pull/24483#pullrequestreview-319132083, would it mean removing existing rbd_op_threads config in current yaml files and moving that into `many_op_threads.yaml` in each directory under qa/suits/rbd?

MahatiC commented:
> Can this PR improve performance?

That's the goal. This would also require aio callback completions dispatched via multiple threads, which isn't supported yet.

MahatiC commented:
@dillaman Also what's the way forward regarding the Objecter/CRUSH crash? Did you get a chance to investigate more or discuss this otherwise?

dillaman commented:
@MahatiC I've been pretty heads-down on trying to get Octopus ready for release this month. There must be a missing lock somewhere that needs to be addressed.

MahatiC commented:
> There must be a missing lock somewhere that needs to be addressed.

That seems like a pretty broad scope. Does it make sense to highlight this issue to the relevant rados team?

dillaman commented:
I am running just your `ImageRequestWQ` commit through teuthology since I want that merged prior to Octopus. As for the librados crash, probably the best way to get attention would be to open a tracker ticket w/ the crash backtrace.

MahatiC commented:
@dillaman Okay. For the librados crash, here's the tracker report, fyi - https://tracker.ceph.com/issues/44311

dillaman commented:
Associated bug fix commits merged in PR #33563

zhengchengyao commented:
> > Can this PR improve performance?
> 
> That's the goal. This would also require aio callback completions dispatched via multiple threads, which isn't supported yet.

I find that boost.asio is on master, I merged this PR base on boost.asio, but performance has not improved.
Do we still need multiple threads of aio callback?  Would you have some good suggestions?
thanks a lot

MahatiC commented:
> I find that boost.asio is on master

Could you refer to this PR?

zhengchengyao commented:
> > I find that boost.asio is on master
> 
> Could you refer to this PR?

yes

MahatiC commented:
@dillaman hi, would these yaml changes be needed in the near future or do I close this for now? I noticed the librbd asio update in #35629 but the librados CRUSH map lookup issue https://tracker.ceph.com/issues/44311 hasn't been resolved yet

dillaman commented:
> @dillaman hi, would these yaml changes be needed in the near future or do I close this for now? I noticed the librbd asio update in #35629 but the librados CRUSH map lookup issue https://tracker.ceph.com/issues/44311 hasn't been resolved yet

I would close for now. Actually, as part of the switch to the new ASIO librados API, the thread limitation for the IO path will be removed (but we will keep a single thread strand for librbd API callbacks and internal non-IO state machines). Once all of that merges, the question then becomes whether or not there is a better librbd API for QEMU to consume that would be higher throughput / lower overhead as compared to the single-threaded AioCompletion callbacks.
